### ISSUE TITLE
SmacPlannerHybrid creates kinematically infeasible path

### ISSUE THREAD

**Issue by tonynajjar** [ros-navigation/navigation2#4271]
## Bug report

**Required Info:**

- Operating System:
  - Ubuntu 22
- ROS2 Version:
  - Humble
- Version or commit hash:
  - It's a fork that tracks commit d509fbf0ee14c15a7692ac8fa852bf9d5821a3e2
- DDS implementation:
  - Fast-RTPS

#### Steps to reproduce issue

I tried it in my custom simulation but I reckon it should also happen on tb3 simulation. You could try to set the global footprint bigger than the local one and try navigating close to a lethal obstacle on the global costmap.

This is the planner/costmap config:

```
global_costmap:
  global_costmap:
    ros__parameters:
      use_sim_time: False
      global_frame: map
      map_topic: /no_global_map_topic
      always_send_full_costmap: True
      rolling_window: True 
      track_unknown_space: False
      publish_frequency: 5.0
      update_frequency: 10.0
      resolution: 0.05
      width: 50 
      height: 50 
      lethal_cost_threshold: 99
      # footprint_padding: 0.5
      robot_base_frame: base_footprint
      footprint: '[ [0.215, 0.42], [0.215, -0.42], [-1.01, -0.42], [-1.01, 0.42] ]'
      plugins:
        - 'static_layer'
        - 'inflation_layer'
      static_layer:
        plugin: "nav2_costmap_2d::StaticLayer"
        map_subscribe_transient_local: True
        map_topic: /mission/geofence/global_costmap
      inflation_layer:
        plugin: 'nav2_costmap_2d::InflationLayer'
        cost_scaling_factor: 1.5
        inflation_radius: 0.7 
        inscribe: false 

planner_server:
  ros__parameters:
    use_sim_time: False
    expected_planner_frequency: 20.0
    planner_plugins: ["GridBased"]
    GridBased:
      plugin: "nav2_smac_planner/SmacPlannerHybrid"
      tolerance: 0.2 
      downsample_costmap: false 
      downsampling_factor: 1 
      allow_unknown: true 
      max_iterations: 1000000
      max_on_approach_iterations: 1000 
      max_planning_time: 1.5 
      motion_model_for_search: "REEDS_SHEPP" 
      cost_travel_multiplier: 2.0 
      angle_quantization_bins: 64 
      analytic_expansion_ratio: 3.5 
      minimum_turning_radius: 0.4 
      reverse_penalty: 2.0 
      change_penalty: 0.15
      non_straight_penalty: 1.50 
      cost_penalty: 5.5 
      lookup_table_size: 10.0 
      cache_obstacle_heuristic: False 
      debug_visualizations: true
      smoother:
        max_iterations: 1000
        w_smooth: 0.3
        w_data: 0.2
        tolerance: 1.0e-10
```

#### Expected behavior

Create kinematically feasible path

#### Actual behavior

Resulting path makes planned footprints collide with global costmap

red thin rectangle = global and local footprint
red thicker rectangle = TEB debug visualization of the infeasible projected footprint
red path = SmacPlannerHybrid global plan

![image](https://github.com/ros-planning/navigation2/assets/14879660/f222908e-fad8-4c01-b548-5c54632e438d)

I also tried with a circular footprint, same result

**Comment by SteveMacenski** [comment]
Questions:
- Can you verify that the costmap being visualized is the global costmap and those footprints are from the global planner & that in the costmap color map that this really is marked as lethal?
-  What happens if you start further from the wall but have the goal near the wall? 
- Have you made any modifications to Smac locally or perhaps missed some syncs which has introduced a bug by branch management? Follow up would be have you tested Nav2 Humble & Rolling and see this in both?
- Is there anything going on with a regular update to the footprint topic (for there to potentially be some race condition issues)?

I've not seen this ever in my testing and deployment of Smac so I don't have an immediate thesis of what's going on beyond asking follow up questions. 

> inscribe: false 

What is this? Are you perhaps using a custom inflation layer?

Not that I think this is the issue, but what if you use the params from the Global Costmap from nav2_bringup + Smac configuration guide params? Perhaps there's some subtle problem that we see with your setup that could help us narrow into it

**Comment by tonynajjar** [comment]
> Can you verify that the costmap being visualized is the global costmap and those footprints are from the global planner

This is for sure the global costmap and footprint.

>  & that in the costmap color map that this really is marked as lethal?

In another test, with circular footprint, I visualized the costmap with the "costmap color scheme" and it shows like this:
![image](https://github.com/ros-planning/navigation2/assets/14879660/795534b6-ba02-4353-a363-ae43572070cc)

Do you have some way of getting the cost of a specific square of the costmap, by e,g, clicking on it to be even more sure?

> What happens if you start further from the wall but have the goal near the wall?

In all cases SMAC ignores that the projected footprint collides. It just throws an error if the goal is exactly at the lethal obstacle

> Have you made any modifications to Smac locally or perhaps missed some syncs which has introduced a bug by branch management? Follow up would be have you tested Nav2 Humble & Rolling and see this in both?

I do have a particular setup yes, it's a fork in between humble and rolling but with no changes to SMAC. Anyway I get it that we should remove this variable by testing on standard Nav2 Humble and/or Rolling. I hit some blockers when trying to do so: 

- I tried to setup rolling with the devcontainer but that's broken
- I tried testing on Humble but for some reason the tb3 demo wouldn't launch properly.
- I was thinking it would be hard to reproduce with TB3's circular footprint but I guess it should be feasible if I make the global footprint bigger.

> What is this? Are you perhaps using a custom inflation layer?

oops, yeah artifact of the fork but I also reverted this to the default and have the same issue.

> but what if you use the params from the Global Costmap from nav2_bringup + Smac configuration guide params? 

Bottom line, I think this has to be done on Rolling to narrow down the issue. Let me know if it's a quick thing you can do if you have a working rolling setup already, otherwise I'll give it another shot soon

**Comment by SteveMacenski** [comment]
>  I visualized the costmap with the "costmap color scheme" and it shows like this

Is the inflation radius you set smaller than the robot's actual minimum radius? I see in the color scheme now that there is no inflation outside of the critically inscribed space. I'm not even sure in the greyscale visualization why that showed a gradient - it looks to be all 253's. Try increasing your inflation radius / decreasing your cost factor to increase the potential field outcome. 

Irregardless, you should have a smoother potential field, it makes Smac behavior better and actually runs faster. 

> Bottom line, I think this has to be done on Rolling to narrow down the issue

Agreed, or at least to the extent that you can test it to check if its an issue there or not to know if its something in the stack itself or something going on on your managed fork. This should be a Planning Server specific issue, so you should be able to use the simple commander API + a launch/params file to bundle up a reproduction example for me to bring up only the planner server and execute the request the triggers this for working the problem. 

> One question, assuming this is a real bug, is there some kind of test that would detect this?

Almost certainly, but I think we need some better understanding of what's actually happening first.

**Comment by tonynajjar** [comment]
So, I spend a lot of time on trying to make it work on Rolling but I'm still not there yet. So I tried on Humble and I think I could reproduce, unfortunately without the planned_footprints as these are not on Humble (and I couldn't backport easily). We can see that a lethal part of the global costmap is in the footprint.

![image](https://github.com/ros-planning/navigation2/assets/14879660/229cc997-0066-4521-9912-e881d70b53f7)

The screen recording is like 8GB so here is a screenshot instead. I made the global footprint a bit bigger than the local one. These are the changes from the default nav2 params:

![image](https://github.com/ros-planning/navigation2/assets/14879660/420a3394-eab7-47ed-8736-936717a3523f)

**Comment by SteveMacenski** [comment]
Can you try what I asked? You still have the inflation way too small - there is no potential field established - and I want to know if that is causing the issue

**Comment by tonynajjar** [comment]
Oops I forgot that part. I'lll try it tomorrow, but assuming it works, we can't just call it a day right? I would think the planner should not generate kinematically infeasible path under any circumstances (throw an error instead). Or am I wrong?

**Comment by SteveMacenski** [comment]
If that's the case, I know the ~3 places I need to look and think about and I can come back with a better answer about "why" and "what" and what a solution would look like, if one is needed. If nothing else, some kind of warning or error or rejection of optimizations should occur in that situation but maybe I missed something. However we don't even know if that'll help, so one step at a time :smile:  But no, that's not the end of it, it just gives me somewhere to start!

**Comment by tonynajjar** [comment]
> I see in the color scheme now that there is no inflation outside of the critically inscribed space. I'm not even sure in the greyscale visualization why that showed a gradient - it looks to be all 253's

sorry for the confusion, this was actually a different configuration


Anyway I think I have the results of what you asked me. robot_radius is 0.3 and inflation_radius is 0.55 but it can still happen. With the default values I couldn't reproduce easily but I don't think it's impossible. 


![image](https://github.com/ros-planning/navigation2/assets/14879660/c6e72525-a1cf-4f3e-9d7c-277c537c86ca)

![image](https://github.com/ros-planning/navigation2/assets/14879660/a5e85599-82ef-4a13-babc-fb4579b99513)



> I don't understand your image though, considering the footprint is in the middle completely disconnected from the path + why the path is orbiting so many times.

To answer this, the circles are not a path orbiting, the red circular line is the global footprint I set and the blue circles are the debug /planned_footprints. (the red rectangle is the local footprint, not important here I guess) In this use case the robot radius is much bigger than the inflation layer, which you disadvised, but I was just testing my assumption that "SmacPlannerHybrid should never plan a kinematically unfeasible path".

**Comment by SteveMacenski** [comment]
What's the black? Something seems a little odd with your setup.

Note that when you use robot radius, it uses center point-checks, not the SE2 footprint. Please use a footprint so that we can have something analog to what you're seeing :-) 

> P.S: The localization of the TB3 demo is very jumpy, it's hard to tell for sure if the plan is unfeasible or if it just "jumped" into the lethal obstacle. I'm trying to get more convincing examples

Note that we clear the robot's current cell, since it has to be clear if the robot is *cough* there. So what you showed isn't odd assuming the very next point in the path isn't in collision. That's to prevent deadlocks so that if you're in collision or just up close to something you can never get out of it. What I'd want to see is collisions mid-path!

**Comment by tonynajjar** [comment]
> Note that we clear the robot's current cell, since it has to be clear if the robot is cough there. 

hmm I'm assuming you're talking about `footprint_clearing_enabled` parameter? I see that's enabled for the obstacle and voxel layer but it doesn't seem to exists for the static layer. I ran a simple experiment to confirm this:

![image](https://github.com/ros-planning/navigation2/assets/14879660/7ffc2c77-345c-49c7-b3c0-6cfd249e8e7d)

the cylindrical pillar is part of the static_layer whereas the rectangular one on the left is not (only part of the obstacle layer). You can see that only the obstacle layer is cleared by the footprint

**Comment by SteveMacenski** [comment]
Lets simplify this, show this happening with your previous environment setup you originally posted with the new configuration - don't try to make it work on the TB3 example / some other situation. Lets focus on the situation you presented. You had showed it happening for many intermediate poses that was unquestionably problematic, show me that again with the new configs so we know if it helped or not. I don't want us to get bogged down in some first/last point details right now - we should be able to see this clear as day as you showed it before with setting A and gone with setting B if we found the problem. 

I think some of what you're using as a proxy (the robot's current localized location) is not a good proxy for planning, so lets remove that estimated proxy and look at the actual results. 

I just grabbed the Smac Hybrid params from the docs and ran it against the same 0.3m robot radius + 0.8 inflation radius on the default Nav2 configs. All seems good to me:

![image](https://github.com/ros-planning/navigation2/assets/14944147/542aa471-5405-4acc-8fb8-c9dd6dc40611)

(not in the right photo that it looks like its colliding; its not, the marker array line thinkness is just set wide. Zooming in its fine.)

When I set the inflation radius to be less than the actual robot radius, it clearly occurs 

![Screenshot from 2024-04-23 15-26-17](https://github.com/ros-planning/navigation2/assets/14944147/ed623b40-2de5-4e7c-acdb-1d9fff39d2d7)

Please show me that you see that too :wink:

**Comment by SteveMacenski** [comment]
So thinking about this logically: if you have a 20m wide robot (for example) but set the inflation radius to 0.1m, then yeah, that seems like a problem because the costmap doesn't even know to *look* for a collision because its all just `0` cost free space, including where you're literally in collision.

So, option:
- Log some complaint that you set the inflation radius less than your actual robot's radius in the inflation layer. Lame, but at least exposes it
- Force the inflation layer's radius to be at least the inscribed radius of the robot and also log the warning to users to tell them that they did it wrong and its not safe. I like this option, along probably others in addition to.
- Insert some checks in the smac planner for this situation and log errors that are missing now for it (maybe good in addition to the bullet above)

**Comment by SteveMacenski** [comment]
First, these are really pretty pictures! The footprint projections are really cool. But anyway -

What is the green? Curious if you know this wasn't an issue before some update or this is always been an issue. If it was a regression, fast forwarding changes until it happens could be a good way to see where it was introduced if new 

This is quite hard for me to debug from afar. What I'm wondering is if there's an issue with collision checking in the analytic expansions. These paths are so trivial that its likely that these are totally formed up of only the analytic expansions without any search. Can you do this again with some more complex request so that search is strictly required at some point and its not trivial to a point that it could be just analytic expansions? That would help answer the following:  Do you consistently see this at the start or end of the paths or both? 

If it only happens at the end (i.e. when analytic expansions are used), that narrows things down. If it happens at the start (uniquely or combined) that also tells us something. 

I cannot replicate this whatsoever, so I feel like this is still a configuration derived issue, but I don't discount a real bug either. However there are a bunch of companies with large non-circular robots using this and it hasn't been reported by them unless its a newly introduced bug. 

>       rolling_window: True

This is still suspicious to me, can we not do that for testing purposes or can you show that it happens with a normal global map without this? Again, collecting data. 

>         inflation_radius: 1.0 ... footprint: '[ [0.215, 0.42], [0.215, -0.42], [-1.01, -0.42], [-1.01, 0.42] ]'

I'm reading this that your robot is 1.01 from the pivot point but your inflation radius is only 1.0. What if you make the inflation radius 2.0 (just sufficiently larger than 1.0)? 

>            cost_travel_multiplier: 2.0

This isn't a param for Hybrid-A*, this is making me suspicious that there is something really odd going on with your params. What if you use the params from https://navigation.ros.org/configuration/packages/smac/configuring-smac-hybrid.html#example exactly?

---

Note: When I say "start" and "end" I don't want to even look at the very first or very last points, just generally collisions are being recorded near the start/end. Lets not miss the forest for the trees.

**Comment by tonynajjar** [comment]
> What is the green?

Green is TEB's path/control sequence

> Curious if you know this wasn't an issue before some update or this is always been an issue.

I can't pinpoint a moment where that "started" happening, I think it was always there

>  Can you do this again with some more complex request so that search is strictly required at some point and its not trivial to a point that it could be just analytic expansions? That would help answer the following: Do you consistently see this at the start or end of the paths or both?

Since the bug mostly occurs when being already close to an obstacle (being in circumscribed area), I would say that it occurs mostly at the start or end of a path i.e. first/last 5 poses. It does not happen when going from free point A to free point B while avoiding an obstacle midway.

> I cannot replicate this on my side so far

That's odd. Do you have all the information needed to try and reproduce it? Important factors:

- robot's start pose is close to a lethal obstacle
- it doesn't always reproduce from the first time - maybe try reproducing exactly the start and end goal that you see in the screenshots
- use the same configuration as me - I also tried with the default configuration and will try again
- Are you testing on rolling or humble? I'm using Nav2 rolling but on Humble ROS system with minor changes just so that it builds (e.g QOS) 

> This is still suspicious to me, can we not do that for testing purposes or can you show that it happens with a normal global map without this? 

Will do. I'll test again with as much "standard" config as possible.

> What if you make the inflation radius 2.0 (just sufficiently larger than 1.0)?

will do

> This isn't a param for Hybrid-A*, this is making me suspicious

that was an oversight, the default config was just copied as is and changed what is needed. Removed

Since you can't reproduce easily, do you have ideas what debug logs I could add inside the code that could give some insight?

**Comment by tonynajjar** [comment]
Let's start with more pictures.

First some successes:

![image](https://github.com/ros-planning/navigation2/assets/14879660/3a46add4-a23f-45fe-b2cb-01516db70947)
![image](https://github.com/ros-planning/navigation2/assets/14879660/0aad33a1-7365-4d5c-9ad5-8fffe44ce435)

the failures:

![image](https://github.com/ros-planning/navigation2/assets/14879660/6efeaf47-3c8c-450c-8aa1-845808771eee)
![image](https://github.com/ros-planning/navigation2/assets/14879660/afe64595-715a-4b67-81bd-720a8c507918)
![image](https://github.com/ros-planning/navigation2/assets/14879660/2e667b7b-2802-490d-9d8a-f89a675d3908)
![image](https://github.com/ros-planning/navigation2/assets/14879660/13126e46-dc9a-48ce-a826-b54694890141)

Default config used:

```
global_costmap:
  global_costmap:
    ros__parameters:
      use_sim_time: True
      global_frame: map
      map_topic: /no_global_map_topic
      always_send_full_costmap: True
      rolling_window: False
      track_unknown_space: True
      publish_frequency: 1.0
      update_frequency: 1.0
      resolution: 0.05
      width: 50 
      height: 50
      robot_base_frame: base_footprint
      footprint: '[ [0.215, 0.42], [0.215, -0.42], [-1.01, -0.42], [-1.01, 0.42] ]'
      plugins:
        - 'static_layer'
        - 'stvl_mixed_layer'
        - 'inflation_layer'
      static_layer:
        plugin: "nav2_costmap_2d::StaticLayer"
        map_subscribe_transient_local: True
        map_topic: /mission/geofence/global_costmap
      stvl_mixed_layer:
        PERFORM_INCLUDE: layers/stvl_cameras.yaml
      inflation_layer:
        plugin: 'nav2_costmap_2d::InflationLayer'
        inflation_radius: 2.0
        cost_scaling_factor: 0.7

planner_server:
  ros__parameters:
    use_sim_time: True
    planner_plugins: ["GridBased"]
    GridBased:
      plugin: "nav2_smac_planner/SmacPlannerHybrid"
      downsample_costmap: false           # whether or not to downsample the map
      downsampling_factor: 1              # multiplier for the resolution of the costmap layer (e.g. 2 on a 5cm costmap would be 10cm)
      tolerance: 0.25                     # dist-to-goal heuristic cost (distance) for valid tolerance endpoints if exact goal cannot be found.
      allow_unknown: true                 # allow traveling in unknown space
      max_iterations: 1000000             # maximum total iterations to search for before failing (in case unreachable), set to -1 to disable
      max_on_approach_iterations: 1000    # Maximum number of iterations after within tolerances to continue to try to find exact solution
      max_planning_time: 5.0              # max time in s for planner to plan, smooth
      motion_model_for_search: "DUBIN"    # Hybrid-A* Dubin, Redds-Shepp
      angle_quantization_bins: 72         # Number of angle bins for search
      analytic_expansion_ratio: 3.5       # The ratio to attempt analytic expansions during search for final approach.
      analytic_expansion_max_length: 3.0  # For Hybrid/Lattice nodes: The maximum length of the analytic expansion to be considered valid to prevent unsafe shortcutting
      analytic_expansion_max_cost: 200.0  # The maximum single cost for any part of an analytic expansion to contain and be valid, except when necessary on approach to goal
      analytic_expansion_max_cost_override: false  #  Whether or not to override the maximum cost setting if within critical distance to goal (ie probably required)
      minimum_turning_radius: 0.40        # minimum turning radius in m of path / vehicle
      reverse_penalty: 2.0                # Penalty to apply if motion is reversing, must be => 1
      change_penalty: 0.0                 # Penalty to apply if motion is changing directions (L to R), must be >= 0
      non_straight_penalty: 1.2           # Penalty to apply if motion is non-straight, must be => 1
      cost_penalty: 2.0                   # Penalty to apply to higher cost areas when adding into the obstacle map dynamic programming distance expansion heuristic. This drives the robot more towards the center of passages. A value between 1.3 - 3.5 is reasonable.
      retrospective_penalty: 0.015
      lookup_table_size: 20.0             # Size of the dubin/reeds-sheep distance window to cache, in meters.
      cache_obstacle_heuristic: false     # Cache the obstacle map dynamic programming distance expansion heuristic between subsiquent replannings of the same goal location. Dramatically speeds up replanning performance (40x) if costmap is largely static.
      debug_visualizations: false         # For Hybrid nodes: Whether to publish expansions on the /expansions topic as an array of poses (the orientation has no meaning) and the path's footprints on the /planned_footprints topic. WARNING: heavy to compute and to display, for debug only as it degrades the performance.
      use_quadratic_cost_penalty: False
      downsample_obstacle_heuristic: True
      allow_primitive_interpolation: False
      smooth_path: True                   # If true, does a simple and quick smoothing post-processing to the path
      debug_visualizations: true
      smoother:
        max_iterations: 1000
        w_smooth: 0.3
        w_data: 0.2
        tolerance: 1.0e-10
        do_refinement: true
        refinement_num: 2
```

as a workaround, is there a config that specifies how far away the `robot_base_frame` or the footprint should stay away from lethal regions?

**Comment by SteveMacenski** [comment]
Hi,

Got it. Huh. Honestly not sure what to say about that off hand. I still can't seem to make that happen. I believe you that there is a problem if that's what you're seeing, but I'm at a loss if I can't reproduce it. Can you make a repository with the costmap, planner configs, launch file, world, and a script to call the server to go from some example A->B that you're using here so that I can try 1:1 what you're doing on my side and use that as a test case to look into? 

If I can poke around, that would be ideal. Otherwise, I'd ask you to look into the collision checker object and log information that you think might help point to a particular issue.

**Comment by tonynajjar** [comment]
Yeah I get that it can be hard to debug if you can't reproduce. Instead of creating a repo from scratch, how about I use what I can from TB3 demo, change to a rectangular footprint + config used above + script and send you my branch?

**Comment by SteveMacenski** [comment]
See the note above I edited, you responded faster than I thought :wink:

I just want something standalone with the planner server / demo script so we can look at this in isolation so we're not messing with the full simulator with localization updates and whatnot that can make the waters a little more murky. Usually what you describe would be fine for me, but this looks to be subtle now (unlike before, which was a clear and obvious problem) so I don't want to leave anything to interpretation

**Comment by tonynajjar** [comment]
I've been delaying creating the reproducible example because it seems like a lot of work. Instead I tried to debug it more myself and I think I have a good lead: by commenting out this [check](https://github.com/open-navigation/navigation2/blob/07fbc50edb75ee97b3650716227caa6925874e84/nav2_smac_planner/src/collision_checker.cpp#L117) and consequently making the collision checker run all the time, I could not reproduce the issue anymore. So I'm guessing footprint_cost_ or possible_collision_cost_ is wrong.

I'm finding it hard "proving" that one or the other is wrong but that's where I'm at right now, if you have any leads, shoot

**Comment by tonynajjar** [comment]
One more insight, by adding this log:

![image](https://github.com/open-navigation/navigation2/assets/14879660/617085bf-6dad-4070-bc66-498fd252e8e7)

I sometimes get a `footprint_cost_` of 0 when it's clearly not 0:

![image](https://github.com/open-navigation/navigation2/assets/14879660/ab743758-1d49-4db8-bcdb-7547a1447568)

**Comment by SteveMacenski** [comment]
> by commenting out this [check](https://github.com/open-navigation/navigation2/blob/07fbc50edb75ee97b3650716227caa6925874e84/nav2_smac_planner/src/collision_checker.cpp#L117) and consequently making the collision checker run all the time, I could not reproduce the issue anymore. So I'm guessing footprint_cost_ or possible_collision_cost_ is wrong.

Not shocked by that I suppose, but unclear. For the second image, there are clearly points in free space, so those `0.0`s might be real. What processor are you running this on? 

I don't have an immediate answer. Some costmap resource not being locked correctly in your fork (or main)? I forget if you're showing this based on your managed fork or Nav2 mainline

**Comment by tonynajjar** [comment]
> Not shocked by that I suppose

It's already useful to know that the problem is likely not in the collision checking itself but in the fact that collision checking is not triggered

> What if this is set to 1.0?

Same thing

> What processor are you running this on?

AMD Ryzen 7 7840HS w/ Radeon 780M Graphics


> there are clearly points in free space, so those 0.0s might be real.

with huge inflation layer still happens: 
![image](https://github.com/open-navigation/navigation2/assets/14879660/db0782bb-ae48-46ea-91b6-1b6e4586b877)


My next step would be to query manually again for the cost at these coordinates and see if it's like you said an issue with locking the costmap. Do you know of an easier way than to create a service that calls `costmap_->getCost`?

**Comment by SteveMacenski** [comment]
Not off hand, but the rviz color scheme seems clear that it is not actually 0. Hence suspecting threading / locking issues getting corrupted data perhaps

**Comment by SteveMacenski** [comment]
Nothing off hand - just general programming experience that issues like data not getting results like they appear to be can be caused by simultaneous read/writes. I'd look at locking the costmap and seeing if that fixes problems, maybe there's an edge case that isn't being covered. Though, if you set the update rate for the global costmap stupid low (i.e. 0.000001 hz) and you still see it, then maybe it isn't that (since there aren't write operations occurring)

**Comment by tonynajjar** [comment]
>  if you set the update rate for the global costmap stupid low

I set it to 0.1 and could still reproduce. Setting it to lower than that blocks the activation of the planner server for some reason

I also suspected `footprint_clearing_enabled` had something to do with it but deactivating it made no difference

**Comment by marcusvinicius178** [comment]
Hi @tonynajjar did you find a solution or new depuring ideia? I am also trying to trick this collision behavior to avoid it, but any combination of parameters config worked https://robotics.stackexchange.com/questions/111025/how-to-configure-nav2s-smac-planner-to-maintain-a-safer-distance-from-obstacles/111044#111044

I have recognized that the inflation layer works better for circular footprint (the blue area inflates correctly, while the rectangle kind of scape from boundaries and seems that don't work). You have mentioned that the collision checker is not triggered right? Is this the main issue cause? I did not dived deep into logging this as you did, but I have ROS2 Galactic and ROS2 Iron installed. Let me know if I can help with some test. 

For me it's strange that all nav2 planners I tested had the same issue while using the Ackermann motion model. Other observation is that months ago when I used a loaded static map, the collision avoidance seemed to work, and now using just rolling windows this avoidance is not working. I could not track what is generating this issue yet,that is frustrating since I  need this working as soon as possible. 

Do you think using some packages from Autoware in my nav2 forked repo to obstacle mapping and path Planning could work?

**Comment by doisyg** [comment]
I did not properly follow the discussion, but if that helps, here is what I use to debug and test the planners: https://github.com/botsandus/planner_playground

**Comment by tonynajjar** [comment]
Thanks a lot @doisyg, this saved me a lot of time.

@SteveMacenski can you try to reproduce:

1- `git clone https://github.com/angsa-robotics/planner_playground -b custom-testing-angsa`
2- `ros2 launch planner_playground test_planner.launch.py`
3- `ros2 run planner_playground send_goal_manually`

I get 
![image](https://github.com/ros-navigation/navigation2/assets/14879660/312f40e8-6d87-41f1-bc53-ebc315e47633)

**Comment by SteveMacenski** [comment]
That doesn't appear to be the same collisions we were talking about - again - please - I want to focus on the main issue of collisions mid-path for a clear and unambiguous situation. The first / last poses have reasons separate of any potential major bug for they they have contact (including clear on start & exact solution requirements) that I don't want to get mixed up with for the main root of a more substantial problem.

Please provide an example for that situation that you've provided many examples previously :-)

**Comment by tonynajjar** [comment]
> Please provide an example for that situation that you've provided many examples previously :-)

hmm, can you send again which examples you're talking about? All the screenshots I provided show a collision at the beginning or end of the path, which is the bug I'm trying to solve. I'm convinced that with a correctly-configured inflation layer, there will not be a collision mid-path (we established that a fix should be create to warn or block a badly configured inflation layer)

**Comment by SteveMacenski** [comment]
A couple where the non-first and non-last points are in collision
- https://github.com/ros-navigation/navigation2/issues/4271#issuecomment-2096307857
- https://github.com/ros-navigation/navigation2/issues/4271#issuecomment-2070326335 (though I think this was fixed once you made your inflation radius not smaller than your robot's minimum radius?)

**Comment by SteveMacenski** [comment]
Oh, ok then. My apologies then, I thought we were still on the original issue. If its just first-last then what you showed me is perfect. This and the issue with the cost scaling factor are on my calendar for Monday to start looking into

**Comment by SteveMacenski** [comment]
Splendid. Not that I need to convince anyone "opensource = good" but another case where @doisyg's one-off testing thing ends up easy to modify and play with for other applications to accelerate them!

**Comment by SteveMacenski** [comment]
@tonynajjar chronicling a bit:

Part 1
- First thing I did was throw out your Smac Planner parameters for the defaults in docs.nav2.org's configuration guide (minus adding `REEDS_SHEPP` since that was required to get it to work) --> still see the issue. This is a good sign to me.
- Increased your `inflation_radius` from the rather exact value to `3.0` to generally just fill the space --> still happens. No surprise there.
- I adjusted your footprint to be the same size but have the center of the polygon be the center of rotation --> still happens. I was kind of hoping that was the issue because I'd know exactly where to analyze in that case
- I set the `tolerance` to `0.0` --> **plan fails to generate** This is what I expected actually w.r.t. the final pose in collision which you example provided is displaying. What you requested is an impossible path. 

So, lets dig into that.

Part 2: Coming to a GitHub thread near you

**Comment by SteveMacenski** [comment]
Please provide a reproduction like you did with the goal and I can look into tha.

**Comment by SteveMacenski** [comment]
@tonynajjar OK, I have a patch for this problem. I don't like how I had to do it, so I'm going to spend some time to make a more elegant solution, but you can count on the final-goal-pose-tacked-on-in-collision problem resolved

**Comment by tonynajjar** [comment]
> Splendid. Not that I need to convince anyone "opensource = good" but another case where @doisyg's one-off testing thing ends up easy to modify and play with for other applications to accelerate them! 

I would suggest bringing this somewhere into Nav2 or the org for long-term maintenance. I wonder what other testbenches would be useful

**Comment by SteveMacenski** [comment]
@doisyg got any new engineers around? Might be a good intro-project to take that and put it into our `/tools` directory as a package for reproducing planning problems.

**Comment by SteveMacenski** [comment]
Part 2:

The key is this line: https://github.com/ros-navigation/navigation2/blob/main/nav2_smac_planner/src/analytic_expansion.cpp#L213 We start at `i=1` since `i=0` is the current node, and we go until `num_intervals - 1`, whereas `num_intervals` is the end point. So we don't actually evaluate the last point's validity to add to the analytic expansion. Then when we set the analytic path that is validated, we do so adding in:

```
  if (goal_node != prev) {
    goal_node->parent = prev;
    cleanNode(goal_node);
    goal_node->visited();
  }
```

which tacks on the final goal pose for exact search in case of some minor deviations due to floating point error somewhere in the expansion. That's how this usually works to give you exact paths and the final goal pose might actually be invalid. I don't think (?) we want to change this mechanic so that you can request a really close goal pose point.

I had a much, much less intelligent solution before while trying to keep analytic expansion working for non-exact valid goals, but I just realized that the analytic expansion really only **should** work when you have an **exact** match. Not checking the final pose and bypassing it seems silly. There is some convenience in having the last point be differently checked (i.e. we check analytic expansion if valid without the last point so you can have something really close to obstacles -- then separately check the last point to see if we should add it or not), but I don't think that's really the intention behind the analytic expansions in the first place.

So... I just changed it to a `<=` statement so we check the last point. Now planning will take a little bit longer in the case that the goal pose is just 1-point into collision, because now the analytic expansion just straight up gets rejected instead of providing a valid expansion except for the final point (that we I could just clip off). It will make those requests slower, buts its probably the **right** solution.

The alternative is that I can keep things as `<` and then I can check the final goal pose separately about whether to add it in (as part of the code block above) so that its can be faster and more close to the intended goal pose with tolerances in this situation. I think that is technically valid and safe, but less intuitive. So, I'm preferring the more strictly-correct method, but I can be easily swayed back if there is an objection

**Comment by doisyg** [comment]
> @doisyg got any new engineers around? Might be a good intro-project to take that and put it into our `/tools` directory as a package for reproducing planning problems. Certainly you or I could do it, but seems like a nice `good-first-issue` style task if anyone came to mind

I don't but if you want me to contribute back, ping me on an issue and I will happily PR

**Comment by marcusvinicius178** [comment]
Hi @SteveMacenski I successfully achieved collision avoidance with the SMAC planner thanks to your recent commit and some adjustments written here: [Params Adjustments](https://robotics.stackexchange.com/questions/111025/how-to-configure-nav2s-smac-planner-to-maintain-a-safer-distance-from-obstacles/111346#111346)

[Avoidance: Ackerman Truck](https://youtu.be/zdDM_fS6fw8)
[Failed to Avoid Obstacle](https://youtu.be/FU8gi_89Z0k)

However, I seek for **earlier collision avoidance to ensure a safer distance from obstacles**. 
Could this be optimized through parameter settings (global or local) or **would it require modifying the SMAC planner source code?** Specifically, I need the transition to steering and collision checking to initiate sooner. Thank you for the recent commit, and I look forward to any guidance or suggestions on this matter!

**Comment by tonynajjar** [comment]
I tested the fix. First of all, no more colliding path with is great:

![image](https://github.com/ros-navigation/navigation2/assets/14879660/cba1bf73-af5b-41ce-ba51-1877ee853cf4)

It looks like two projected footprints were removed and not just the one in collision. I don't know if it's a big deal but might hint to an issue in the fix (?)

before:
![image](https://github.com/ros-navigation/navigation2/assets/14879660/3cadf0e9-094d-41d7-92ec-c6b2a12ee6ce)

after:
![image](https://github.com/ros-navigation/navigation2/assets/14879660/ca1de94e-2529-4531-bc67-a9cfaab5938b)

> Please provide a reproduction like you did with the goal one and I can look into that.

This is next up for me

**Comment by SteveMacenski** [comment]
> EDIT: Looking at it again, I doubt it's an issue because it looks like a completely different path was selected, as opposed to clipping two final poses from the original path. Will leave my comment there anyway for you to be the judge.

Correct. Because this is no longer finding the exact analytic solution, its finding another solution within your specified tolerance parameter.

**Comment by tonynajjar** [comment]
> > This is next up for me
> 
> It didn't take long, @SteveMacenski try it with: `git clone https://github.com/angsa-robotics/planner_playground -b reproduce-collision-at-the-start`

@SteveMacenski any estimate when you'll be able to take a look at this?

**Comment by SteveMacenski** [comment]
The Jazzy release stuff has put this as a secondary problem for me until thats all squared away for a release. I hoped to get to this over the weekend but that didn't pan out. I expect that this is a much simpler problem though if you wanted to take a look. You shouldn't have to go too deep since we convert the start to a grid-space early on and then collision check it. I expect we have some clearing under start that we say is "OK" that could be made into an optio

**Comment by SteveMacenski** [comment]
@tonynajjar this is easily explained. 

```
  // Note: We do not check the if the start is valid because it is cleared
  clearStart();
```

We clear the start since the robot is there, it is thus "clear" because we are physically there already. It must be clear (though perhaps in contact with something) if the robot is in this location. When I add in while removing `clearStart()`: 

```
  // Check if starting point is valid
  if (!_start->isNodeValid(_traverse_unknown, _collision_checker)) {
    throw std::runtime_error("Starting point in lethal space! Cannot create feasible plan.");
  }
```

It triggers that there is a collision and refuses to plan. This is the behavior we have in Humble & Iron but not main. I think this is a sensible behavior to have, don't you? Else you could end up in a forever loop of not being able to move when in collision with something in the static map that moved and thus is no longer actually there (or if in a tight space cannot get out). 

I suppose we could make this an option about if you want to clear it or check it, that would be straight forward. I however don't see much reason you'd want to be so strict about it unless you were generating starting point paths where the robot physically was not correctly located for testing (or I suppose precomputation). 

Let me know what you think. Its an easy resolution either way (closing this as finally completed; and/or adding in the option)

**Comment by tonynajjar** [comment]
That's my bad, I didn't mean for the starting pose to be in collision. The problem I'm trying to reproduce is when the collision is at some pose at the beginning (but not the first pose). I was able to reproduce and colored the first pose in red so you can be sure that it's not the first pose that's in collision:

![image](https://github.com/ros-navigation/navigation2/assets/14879660/ec04697d-42f7-4e4d-a14c-9665aaa45e78)

I updated the branch `reproduce-collision-at-the-start` so you can pull and try again

**Comment by tonynajjar** [comment]
Any chance this could be looked at this week again? (sorry to pressure but deployment is soon and that's still an issue :see_no_evil:)

**Comment by SteveMacenski** [comment]
@tonynajjar I'm going to look into it this afternoon, but I think even before starting to program I know what it is. I've been having this on the back of my mind and I guess my brain was crunching on it subconsciously.  

I think what's happening is that we set the starting cell to clear. So, if you also happen to rotate in place within the same cell, that would also be free at any orientation, not just the starting pose. And that's what it looks like is happening with your example. Its easy for me (once this compiles and I come back from lunch) to test if that's the case with the same experiment I did above removing `clearStart`. 

Then, I'll just need to change how we handle the starting point as "special" rather than editing the costmap for how we process it. Its a little unclear to me how it is that we have a virtual rotate-in-place for hybrid-A* at the start, since that's not a behavioral option. But, that is a behavioral option for State Lattice so that's a bug nonetheless

**Comment by SteveMacenski** [comment]
Yup, just removing the `clearStart` fixes this. I'll just find a workaround for how to handle the first starting pose differently :+1: 

![image](https://github.com/ros-navigation/navigation2/assets/14944147/4f561ad2-ffac-44c9-947c-97ff1983c141)

**Comment by SteveMacenski** [comment]
https://github.com/ros-navigation/navigation2/pull/4418

I think this squares up this ticket :-) 

Thanks for filing this @tonynajjar and these are good details to have worked out. The start pose issue that we just handled is actually unique to `main` so there's no need to backport to iron/humble, those just reject requests where the start is occupied straight-up so there's no problem for existing binary users.

**Comment by tonynajjar** [comment]
Cool! Will try it out tomorrow. I think there is one more thing to do in this issue - warn or block if the inflation layer is not configured big enough. We discussed it in the beginning

**Comment by SteveMacenski** [comment]
I thought that was something you were going to open a PR for :upside_down_face:

**Comment by tonynajjar** [comment]
I wanted to gather all the TODOs for this issue and tackle them at once but that's not how it panned out 😅 I'll reopen just in case my short-term memory fails me until I create another ticket (or PR) for it tmrw

**Comment by SteveMacenski** [comment]
Add a ERROR log instead of a ticket and that should suffice! At least that way someone looking through logs to debug would see that they made a mistake even if we don't actively prevent them from making that error.

**Comment by marcusvinicius178** [comment]
Hi @SteveMacenski @tonynajjar @doisyg  I am using ROS2 Iron, the clearStart fix is not needed as you have mentioned. However I am still facing collision issues.

I have already increased the radius_inflation parameter either on local and global costmaps to be greater than robot's footprint and set the resolution of costmaps from small values (0.01) to huge values ( 10.0). 
Unfortunately the result is always the same: A) Collision against obstacle or B) Trucking getting stuck in front of my obstacle reattempt several new paths , as you can check on this [Video](https://youtu.be/I6ofjPSHCXc) and on the pictures below

Can you please guys assist me to fix this? 

My nav2 parameters are in this link: [truck_nav2_params.yaml](https://pastebin.com/7F1jDp1g)
And issue is described also here: [collision with smac planner](https://robotics.stackexchange.com/questions/111025/how-to-configure-nav2s-smac-planner-to-maintain-a-safer-distance-from-obstacles/111346#111346)

![Screenshot from 2024-06-20 13-17-04](https://github.com/ros-navigation/navigation2/assets/43826438/890081c6-d517-47a8-9f2f-ad7f05ff8bb3)
![Screenshot from 2024-06-20 13-16-46](https://github.com/ros-navigation/navigation2/assets/43826438/4311ca8f-da2a-4242-9ab4-fca306a92153)
![Screenshot from 2024-06-20 13-14-16](https://github.com/ros-navigation/navigation2/assets/43826438/6e724fb1-5d24-4af2-803c-98794377376c)
![Screenshot from 2024-06-20 13-13-55](https://github.com/ros-navigation/navigation2/assets/43826438/97993363-e5c3-4d48-8a91-15adf5ec1653)
![Screenshot from 2024-06-20 13-13-47](https://github.com/ros-navigation/navigation2/assets/43826438/c96eb1d2-211f-462c-b7db-7a2f112c65ad)
![Screenshot from 2024-06-20 13-13-14](https://github.com/ros-navigation/navigation2/assets/43826438/e01f573a-1298-4a3c-a134-b69cf004fdd2)
![Screenshot from 2024-06-20 13-12-46](https://github.com/ros-navigation/navigation2/assets/43826438/876d67ae-a765-463f-a992-e243ccfa652d)
![Screenshot from 2024-06-20 13-12-23](https://github.com/ros-navigation/navigation2/assets/43826438/24f1f7e6-5d47-4139-a8e9-dd8cbb5f0943)

**Comment by marcusvinicius178** [comment]
@tonynajjar I was busy while you were fixing the issue, I am reworking the planner tasks now...I've missed this part, I will check how to use this tool you have mentioned and then I can submit the results here. One question for now, In your case, you don't need to avoid an obstacle in the middle of your trajectory like me?

**Comment by tonynajjar** [comment]
I do, but that was never a problem if the inflation layer is configured correctly, only collisions at the start and the end

**Comment by marcusvinicius178** [comment]
> I do, but that was never a problem if the inflation layer is configured correctly, only collisions at the start and the end

Oh, that’s concerning 😬. Could I take a look at **your configuration file and get the dimensions of your robot** (footprint or radius)? For reference, my truck is 10 meters in length and 4 meters in width. I’ve set the inflation layer values to be greater than these dimensions, even using a value of 50.0. However, I’ve noticed that the obstacle layers seem to have a numerical value threshold. Visually, the inflation seems to stop increasing after setting a value around 5.0.
Curious to see how you managed this avoidance issue (or this is a specific issue from my side having a huge robot hehe).

**Comment by tonynajjar** [comment]
You can get my planner config from my fork of the planner_playground I shared a few comments up

**Comment by iradionov** [comment]
Hi, guys!
The problem seems not connected with the inflation radius value (it doesn't guarantee getting a footprint collision-free trajectory). I guess the problem is located in [this](https://github.com/ros-navigation/navigation2/blob/main/nav2_smac_planner/src/collision_checker.cpp#L117) condition. We compare the [shifted](https://github.com/ros-navigation/navigation2/blob/main/nav2_smac_planner/src/collision_checker.cpp#L115) (?) point cost with possible_collision_cost_ value and most of the time the method footprintCost(...) is not executed. 

The same problem is discussed [here](https://github.com/ros-navigation/navigation2/issues/4361) and [here](https://robotics.stackexchange.com/questions/109722/mppismacplannerhybrid-navigation-with-a-large-base-footprint-problem/111825#111825)

Here is how the default behavior looks like:
![default planner](https://github.com/ros-navigation/navigation2/assets/45877502/f788a803-30a8-4f10-997a-7c47b868b03f)

Here is the result with [this](https://github.com/iradionov/navigation2) modification:
![modified planner]](https://github.com/ros-navigation/navigation2/assets/45877502/116598b4-ffea-4ba6-8ccd-697de236af91)

**Comment by SteveMacenski** [comment]
That simply removes the optimization. The issue is that your inflation layer is not properly setup as discussed at length in the ticket above. If using `main` / `rolling`, you should see an error message about this now: https://github.com/ros-navigation/navigation2/pull/4423/files -- though that covers in the inscribed radius.

What happens if you set your inflation radius to be larger than your **inscribed radius**? The tuning guide, ROSCon talks, and example configurations all highly suggest that you use a smoother potential field for performance and path quality improvements. While actual collisions shouldn't happen, this is why most people don't see this issue since having a tiny inflation with pure 0-cost space is a bad configuration for those algorithms.

**Comment by iradionov** [comment]
Thank you for your prompt response.
There wasn't an error message because the inflation layer radius is larger than the inscribed radius. 
I understand the idea of compute time optimization but extending an inflation layer radius leads to increasing costs of this region. 
So instead of getting trajectories like this 
![sample1](https://github.com/user-attachments/assets/a6d57c32-a5a7-4f16-89a4-8fe5e6bbe1eb)
we increase the inflation layer radius up to the circumscribed radius and can observe something like
![sample2](https://github.com/user-attachments/assets/6bfcfdfe-fba0-43b5-9a4b-e9d7fd8c6ea6)
In my case (coverage task) this is not acceptable, but is acceptable for most tasks.

The last question:  why do we get the cost of the [shifted point](https://github.com/ros-navigation/navigation2/blob/main/nav2_smac_planner/src/collision_checker.cpp#L115)?

**Comment by SteveMacenski** [comment]
> The last question: why do we get the cost of the [shifted point](https://github.com/ros-navigation/navigation2/blob/main/nav2_smac_planner/src/collision_checker.cpp#L115)?

It has to do with the fact that casting it to an `unsigned int` floors the value (example: `4.9` will become `4u`). Adding 0.5 makes it so that we round to the nearest whole number when down casting (in our example now `4.9` would be come `5u`) 

>  we increase the inflation layer radius up to the circumscribed radius and can observe something like

That is a tuning thing with the `cost_penalty`. You could set that low so that there is no incentive to stay away from higher cost areas (outside of collision, which will always be rejected) :-)

**Comment by tonynajjar** [comment]
Unfortunately I can reproduce this issue on Nav2's Jazzy again

![Image](https://github.com/user-attachments/assets/f564a66a-5b18-4c51-8a9e-1497eaac99fa)

@SteveMacenski would you have the time to take a look? You can use the same instructions as [here](https://github.com/ros-navigation/navigation2/issues/4271#issuecomment-2112586038), I updated the branch with the new start and goal.

**Comment by SteveMacenski** [comment]
I'm not entirely sure what I'm looking at in that polygon. Is that the robot's current pose polygon? Do you have footprint clearing enabled in your costmap configs to make that pass as intrinsically navigable?

**Comment by tonynajjar** [comment]
I created this example and pushed it to the branch `#4271` if you want to reproduce

>  Is that the robot's current pose polygon?

The start pose is the green arrow on the top right and the goal is on the bottom left.

![Image](https://github.com/user-attachments/assets/e7c280e1-82ee-44fe-8c4c-f5d5817dfdab)

The blue footprints are SMAC's `/planned_footprints`

![Image](https://github.com/user-attachments/assets/3a38ce47-0000-458e-9660-73656d504763)


> Do you have footprint clearing enabled in your costmap configs to make that pass as intrinsically navigable?
nope, you can find all the config [here](https://github.com/angsa-robotics/planner_playground/blob/%234271/config/nav2.yaml)

**Comment by SteveMacenski** [comment]
@tonynajjar this seems to relate to the PR that we closed this under: https://github.com/ros-navigation/navigation2/pull/4423

Your image looks like the center point of the robot is in zero costed space (the footprint center isn't in the center, right?). From the optimization in the Smac Planner to only collision check the full footprints, we check [what the maximum cost in which it is possible for the robot to be in collision](https://github.com/ros-navigation/navigation2/blob/main/nav2_smac_planner/src/smac_planner_hybrid.cpp#L262). If its zero, then [you should be seeing a log message about it](https://github.com/ros-navigation/navigation2/blob/main/nav2_smac_planner/src/collision_checker.cpp#L58-L64). Check these values and logs please. 

So, you would need to have the inflation radius of non-zero cost set at least the circumscribed radius of the robot for that optimization to properly do its job. The [collision checker](https://github.com/ros-navigation/navigation2/blob/main/nav2_smac_planner/src/collision_checker.cpp) should be able to handle the situations when `possible_collision_cost_ == 0.0f` and simply just do the full collision check anyway (more computationally expensive, by the way), but maybe there's an edge case we missed or something isn't getting properly translated?

**Comment by SteveMacenski** [comment]
Ah. Here's the issue.

We find the cost for the circumscribed radius here:  https://github.com/ros-navigation/navigation2/blob/6a51f1ac33ccbe980e2c4efb2ab09c49d2515d34/nav2_smac_planner/include/nav2_smac_planner/utils.hpp#L85. That inflation layer function is implemented here: https://github.com/ros-navigation/navigation2/blob/6a51f1ac33ccbe980e2c4efb2ab09c49d2515d34/nav2_costmap_2d/include/nav2_costmap_2d/inflation_layer.hpp#L149

See something missing? There's no respect for the inflation radius threshold. We can update `findCircumscribedCost` to check if the `circum_radius > inflation_radius` and if so set the cost to `0.0`. 

That'll fix things so that the collision checking properly works, but will then just collision check all the full footprints - which is slow. I think the warning addition is needed too to make sure folks have good experiences. So I'd add that as well!

(We should update [MPPI's](https://github.com/ros-navigation/navigation2/blob/6a51f1ac33ccbe980e2c4efb2ab09c49d2515d34/nav2_smac_planner/include/nav2_smac_planner/utils.hpp#L85) too)

Can you open a PR>

**Comment by tonynajjar** [comment]
> We can update findCircumscribedCost to check if the circum_radius > inflation_radius and if so set the cost to 0.0.

To be clear, so like the screenshot I sent right (except return 0.0 and not -1)?

**Comment by SteveMacenski** [comment]
I believe either would work - yes. I think maybe 0.0 would be better just for clarity. I think that should be ERROR not WARN, but otherwise yeah!

**Comment by tonynajjar** [comment]
The chronicles continue: I have another collision case which I believe is not due to a wrong inflation radius 

![Image](https://github.com/user-attachments/assets/6680bc87-c8ab-4a8f-80d8-d8b4c0c50c93)

Same image but zoomed in:
![Image](https://github.com/user-attachments/assets/93124031-fe5b-406c-bb8b-3e7a3a4fcd97)


I noticed that reducing the `reverse_penalty` from `5.0` to `3.0` yields this feasible path instead but nevertheless a higher `reverse_penalty` should not be a reason for planning into a lethal space:

![Image](https://github.com/user-attachments/assets/a7f78f86-d1ab-43c6-b8b1-f8d08dcc1c14)

Changing `analytic_expansion_max_length` to both small and big values does not fix the issue

As usual, it's reproduceable in my [planner playground fork](https://github.com/angsa-robotics/planner_playground/tree/%234271)

Any ideas what else could be the reason?

**Comment by tonynajjar** [comment]
Update: I'll give more details when I'm on my computer again but I'm pretty sure it's because of the smoother. i see there is a feasibility check in there but didnt get the time to understand the issue there yet

**Comment by tonynajjar** [comment]
> Update: I'll give more details when I'm on my computer again but I'm pretty sure it's because of the SMAC smoother. i see there is a feasibility check in there but didnt get the time to understand the issue there yet


https://github.com/ros-navigation/navigation2/blob/711817f995d5a33dba29e7330874bd93b50fe860/nav2_smac_planner/src/smoother.cpp#L170

In the smoother we check if the cell cost is greater than MAX_NON_OBSTACLE to reject the smoothed path -> what if `possible_collision_cost_` < cell cost < `MAX_NON_OBSTACLE`? Then the path could be non-feasible depending on the orientation.
I assume that's the problem, what do you think?

FYI 

smoothed planned footprints:

![Image](https://github.com/user-attachments/assets/1fb19d0f-7b82-4709-ba12-8249d4d9a110)

unsmoothed planned footprints:

![Image](https://github.com/user-attachments/assets/dfc68e63-c387-485f-b0d5-083318b27891)

**Comment by SteveMacenski** [comment]
It is true that the smoother doesn't not take into account the footprint because it itself doesn't take into account the orientation in order to transform that footprint to an orientation to SE2 collision check. It is a very simple post-processing smoother just to take the edge off of little wiggles or harsh turns that can be smoothed out for smoother path quality.

I'd be open to suggestions as to how to improve it. Perhaps have a check at the very end considering the new orientations with the footprint and reject the smoother update entirely if not valid. Perhaps going a step further within the smoothing loop to iterate through each point - keep what we do now for collision checking obvious problems - and at the end of iteration (https://github.com/ros-navigation/navigation2/blob/711817f995d5a33dba29e7330874bd93b50fe860/nav2_smac_planner/src/smoother.cpp#L182) we `updateApproximatePathOrientations` and do a full SE2 check on the path. If invalid, then we do what the other failure conditions do and `path = last_path; --> return;`.

**Comment by tonynajjar** [comment]
small update, I think my collision issue is on the controller side (it has its own collision-checking and doesn't blindly follow the global path) so I'll have to invest time on that front first

**Comment by SteveMacenski** [comment]
Can we close this then in the interim? Happy to have you reopen it again if we're back in Smac-land

**Comment by tonynajjar** [comment]
I mean it's still an issue to keep track of. I would love if someone picks it up and that's more likely if it stays open but not a strong opinion.

**Comment by tonynajjar** [comment]
I think I found yet another scenario where SMACHybrid plans right into lethal space and the cause is not one of the previously mentioned reasons (smoothing or inflation radius) -> I have smoothing set to False and my inflation radius is big enough.


I found it to collide when the starting pose is close to a small number of lethal cells.

A few comments about the following pics:
 - Nevermind the black donut, it plays no role
 - green line is unsmoothed plan and blue rectangles are the unsmoothed footprint
 - the TF is base_footprint

Scenario 1: Close to multiple lethal cells - Fine ✅ 

![Image](https://github.com/user-attachments/assets/81b589a9-c51a-4f8f-a9bf-abe8b19c961c)

Scenario 2: Has some distance from a couple lethal cells - Fine ✅ 

![Image](https://github.com/user-attachments/assets/7bd29148-deaf-4bf7-a171-41dd145ed4b0)

Scenario 3: Very close to a couple lethal cells - Collision ❌ 

![Image](https://github.com/user-attachments/assets/f23259a5-f71c-4dee-91be-6225757cec74)


Any idea what could be causing that? As always, would be happy to provide a reproduction if you will have time to look into it yourself.

**Comment by tonynajjar** [comment]
small update: I noticed that reducing the `reverse_penalty` generates a feasible path but it's not really a feasible solution for me because I don't have perception in the back - I want to reverse only as much as necessary so my `reverse_penalty` needs to stay relatively high

![Image](https://github.com/user-attachments/assets/5fb8d5f7-507e-4f35-b496-fef373e0d07d)

**Comment by SteveMacenski** [comment]
Interesting - I have noticed in your plans in general that you have a pretty big jump in orientation at times. It seems related to the small turning radius you use, but that results in a pretty massive jump from bin-to-bin in the rotations conducted. That was not anticipated in the original paper or implementation to have such small turning radius' with respect to your costmap resolution. Default Nav2 uses `0.4m` turning radius with a 5cm costmap resolution, so that's an 1:8 proportion. The Hybrid-A* paper is something on the other of a few meters for the turning radius at a 5cm resolution (>> 1:10).

For example, when the turning radius is sufficiently small relative to the costmap resolution, to leave a quantization (the costmap cell), it will need to have an angular change that is quite large (i.e. arc of the circle is large to move `sqrt(2)` cells). Doing so means you have a very abrupt change that covers a great deal of angular space. That's less an issue specific to Smac Planner but all A* planners - your search discretization relative to the world model is an important feature.

So when you make those large angular jumps and you have - for example here - a tiny point obstacle, you can effectively jump straight over it.  Since the collision checking is raycasting on the footprint's polygon it isn't checking the internal space. If you look at the planned footprints, only touches the lethal space and it is clipping a tiny fraction of the cell. Unless we had the raycasting resolution set to 0.001mm, we wouldn't reliably catch that. This isn't unique to Smac or even Nav2, all planners would have this issue, essentially. 

So what can be done here:

- Increase your turning radius so you don't have these discontinuous jumps. Frankly I find it hard to believe this path below (ignoring the collision) is really what you want your robot to do. I think that's the best answer that addresses the underlying issue. 
- Decrease the costmap resolution to be lower proportionately to your selected turning radius if you can't change the turning radius
- Implement a feature to add a 'flood-fill' collision checking option for the footprints (though will come at a computational cost). This is the second best answer. I suppose we could make a check that if any point is not just possibly in collision due to the circumscribed radius, but also (based on a parameter) if any point on the footprint is found in the `253` inflated space that we break out of the perimeter check and use flood fill. 
- Parameterize and allow you to increase the collision checking raycasting resolution (which is easy). However, I think using some insanely fine resolution that would catch tiny, almost inperceptable clips like that would be so low that you'd probably just be better off flood filling for the size of your footprint. But you can easily calculate the number of checks by `perimeter * check resolution > size_footprint_x_cells * size_footprint_y_cells`   
- ... probably some other changes in Smac, but those are the most obvious targets. We either need to change how collision detection works or increase finer discretizations in search resolution by parameters (turning rad / costmap resolution) or in the core method itself 
- Don't enforce such a high reversing penalty to let it get out of that space
- Set a behavior near an obstacle to have the robot back up a bit before trying to plan

See the difference in volumetric sweeps in 

![Image](https://github.com/user-attachments/assets/4bb3be71-e7ea-4978-a130-fb76fe12aac0)

vs 

![Image](https://github.com/user-attachments/assets/a04f86ce-ed45-47af-a653-cb09dcefaac4)

**Comment by tonynajjar** [comment]
> Increase your turning radius so you don't have these discontinuous jumps. Frankly I find it hard to believe this path below (ignoring the collision) is really what you want your robot to do. I think that's the best answer that addresses the underlying issue.

> Decrease the costmap resolution to be lower proportionately to your selected turning radius if you can't change the turning radius

I tried but that doesn't seem to resolve it. This is with a turning radius of 0.4 and a costmap resolution of 0.05. Take a look at this following case: the footprint vertex outright intersects with the occupied cells close to the starting position - I can't explain or understand why. Then there is a second occupied cell which is fully within the footprint - this is explained by the fact that `footprintCost` only considers the edges for collision checking

![Image](https://github.com/user-attachments/assets/beb4fce6-5ad1-4e46-9d93-d3cc8b590066)

by the way these are the unsmoothed path, not to add any more potential sources of problems in the collision checking

> Don't enforce such a high reversing penalty to let it get out of that space

I don't mind it going back a little bit but I don't think I can successfully tune `reverse_penalty` to achieve that (without sometimes planning for long distances backward)

So far I'm seeing the flood-fill collision checking as the only viable option for me

**Comment by SteveMacenski** [comment]
@tonynajjar can you give me a planner_playground configuration / start/goal pose for this case you show above? This is a good one that I can look into.

If you don't want to wait on me, something that I'd try is to increase the raycasting resolution, it could be the same thing I mentioned before that since this is just a point obstacle it is missed in the raycasting resolution which is set for every 1x cell length. It could be really unlucky that this isn't marched across, which maybe we should increase that resolution to every 0.5x cell lengths instead. For testing purposes I'd make it 0.1x cell lengths just to make sure there's no way we missed it. 

I'd also add some prints to make sure that it is checking the full SE2 footprint and isn't for some reason only checking the center point cost. 

These single-point obstacles are difficult to handle it seems except if we did:
- flood fill collision checking the footprint's internal space
- swept the area from the root footprint from an expansion to the child footprint and checked those area points in addition to the outline poses 

Depending on the size and shape of the footprint (number of checks on perimeter > number of checks flood filled?)  which would be more efficient would change. I generally think swept area + the existing outline check would make sense. Basically the logic would go:

- Check the perimeter footprint as we do now
- If the maximum cost is lower than 253, use that value
- Else, find the intersection of the polygons for the "new" space swept by the expansion and check its fully internal cells  (if parameterized setting is on). Luckily with how the Smac Planner is architected, that should involve only modifications of the Collision Checker which is isolated from the rest of the code and applies to then all of the planners. 

I think that would outright solve your issue in all its forms, assuming what I'm saying is correct and there isn't a bug elsewhere. The first partially-intersecting point is why I ask for the planner playground to see why exactly that one is missed (bug or unlucky raycasting). But, if we swept area, if the perimeter check missed it, then the next expansion from it would have the point fully encapsulated by the footprint's area and fail that check. Thus, no plan would contain that branch since any motion from that footprint is invalid. We could of course increase the check resolution, but the swept area is a more complete solution that resolves the other discretization issue. 

Is that something you're interested in contributing or should I put this on my work queue?

The hardest part is the polygon overlap bit. This can be prototyped with boost geometry pretty easily using `bg::difference(poly2, poly1, diff);` (and maybe we just keep boost as a Nav2 dependency?). Though, ideally we remove boost dependencies to keep the binary install size down and that's something that Generative AI can provide for us I think as long as we create manual test cases we validate on paper. I suppose there are other libraries we can pull from, some which are header only which don't increase the install size as much like https://github.com/AngusJohnson/Clipper2. I'm also not known for optimizing for image sizes so if boost works and you can't justify the time to replace it with something else, then I'm perfectly fine merging that and filing a ticket to let folks know there's the opportunity if they care to replace it themselves in a PR.

**Comment by tonynajjar** [comment]
> can you give me a planner_playground configuration

1- git clone https://github.com/angsa-robotics/planner_playground -b 4271 and build
2- ros2 launch planner_playground test_planner.launch.py
3- You can move the footprint around with the black donut (interactive marker) so place it similar to the screenshots and send a goal with 2D Goal Pose

let me know if something doesn't work, I modified quite a bit.


>  something that I'd try is to increase the raycasting resolution, it could be the same thing I mentioned before that since this is just a point obstacle it is missed in the raycasting resolution which is set for every 1x cell length. It could be really unlucky that this isn't marched across, which maybe we should increase that resolution to every 0.5x cell lengths instead. For testing purposes I'd make it 0.1x cell lengths just to make sure there's no way we missed it.

If I got you right I tried it like that:

![Image](https://github.com/user-attachments/assets/a5c786cb-cd2a-4f24-aed0-eec43b8f4f31)

but it didn't help:

![Image](https://github.com/user-attachments/assets/caabf4f5-f46e-4a0d-b7e9-86f8ca1edf32)


**What did work on the other hand**, is a `fullFootprintCost` implementation with the help of an LLM (cells intersecting vertices + inside cells). It is about 10 to 20 times slower than `footprintCost` and that's for a relatively small footprint and I even specialized it for a rectangular footprint. This is probably what we'll go with on the short-term since speed is not an issue for us but for Nav2 we need a more generic and preferably faster solution I suppose. 

I'll look into the swept area method you mentioned a bit

> Is that something you're interested in contributing or should I put this on my work queue?

The sad part about finding a quick workaround is that I can't justify spending more time on a cleaner solution 😢 but I'll try for the next couple of hours and see if I get somewhere with it

P.S:  By the way, I never mentioned the context of such small obstacles if it's interesting for you; it's basically false positives from our perception e.g. uneven ground. I think that's pretty common so I'm surprised that issue was never reported in the past

**Comment by tonynajjar** [comment]
If I understand correctly we have 2 implementation options:

1. Get diff between 2 consecutive footprints + check cost of diff cells
2. Check all cells in footprint

Some thoughts:

- For strategy 1, would you just check only the diff or also the interpolation between the 2 footprints? As you mentioned earlier this is especially relevant for rotating motions when the minimum turning radius is small because a lot of distance is covered. 
-  I assume the reason why you mentioned strategy 1 is also to save on computation. I think the winner in performance depends on the footprint shape and size; getting the diff might outweigh the benefit of checking fewer cells

**Comment by tonynajjar** [comment]
In the short time I had, I drafted a PR for strategy 2: https://github.com/ros-navigation/navigation2/pull/5250/files
I can confirm that it fixes the issue:

![Image](https://github.com/user-attachments/assets/3458a452-fe02-4a13-bcd4-04c6cefee76f)

![Image](https://github.com/user-attachments/assets/37df4905-4727-40bd-9cc5-7fbdb70ade35)

Waiting for your thoughts before going further. I would have time to continue on this track and finish up the PR if it's viable. I think there is a lot of room for optimizations but we could merge this PR relatively soon as a "bug fix" and follow up with optimizations later

**Comment by A-I-Ivanov** [comment]
Hi Tony&Steve, 
I'm trying to follow this bug/issue since I'm attempting to use SMAC with a robot that is basically diff drive but with an extended rectangular footprint (i.e. it can and sometimes must turn [nearly] in place). The key thing I'm trying to understand here is the collision checking issue that has been brought up. How exactly is this being done? I read up on the documentation stating that the lattice based collision checking is done by considering different points along a path, but this seems to break down along small turning radii. That is assuming the points are sampled based on arc-length or along track distance. With a small turning radius the path length can be short while the angular distance can be quite large. Assuming my understanding of the collision checking is correct, can we test utilizing a different collision sampling method that includes angular (not just along track) distance? 

Apologies if my questions are out of bounds since I'm relatively new to using SMAC and don't understand the details yet.

**Comment by SteveMacenski** [comment]
> If I got you right I tried it like that: ... but it didn't help

Huh, do you see that it checked that cell? If that was implemented properly, that should be checking 10x per cell length which would absolutely cover that. 

Something I'd check also is the coordinates, they're `int`. Maybe the rounding error of the cell conversions puts that footprint line segment at a slightly different angle that it misses the single-point collision. Maybe we can keep things as map continuous coordinates for as long as possible before the final conversion into the discrete cell for checking. 

I think both of these are best validated / checked in a visualization. Publish the poses checked in the line cost check and then the line segment itself being used. 

> I can't justify spending more time on a cleaner solution

I understand, even dirty solutions are good starting points that I/we/someone can clean up at a later time - I appreciate the testing and reports either way!

> For strategy 1, would you just check only the diff or also the interpolation between the 2 footprints? As you mentioned earlier this is especially relevant for rotating motions when the minimum turning radius is small because a lot of distance is covered.

I suppose we could do either:
- Use the 2 polygons to statically identify the area that the new footprint encapsulates that the old footprint didn't (which might be in multiple pockets, not promise it'll be a single polygon output). Check the internals of those polygons (not the border, the border is handled by the perimeter check).
- We use the original polygon -> new polygon to rotate the footprint at a higher resolution and check each perimeter check at that fine resolution **or** during this higher resolution primitive check, we create a `std::set` of cells so we remove all duplicate checks (!). This should actually cover the "sweep" areas that the original and new polygons don't cover when you have a long robot with a corner and is considerably easier to implement IMO.

I think we should do that second bullet point. It also has the benefit that we can add in the original polygon outline check to that set too so we never check duplicate cells, ever. It would be slower, but I don't think it would be 20x slower as you mention

> I assume the reason why you mentioned strategy 1 is also to save on computation. I think the winner in performance depends on the footprint shape and size; getting the diff might outweigh the benefit of checking fewer cells

If we had a robot who's area would involve 100 checks and a perimeter that would involve 100 checks, then it wouldn't really matter. I suggest that the choice of perimeter vs area checks depends on the shape, but more shapes are going to be more efficient with perimeter check than a full area check. So, if we stick with the perimeter check, then if we need to do some kind of area check, then I'd like to do so involving as few cells as possible -- hence only area checking the diff area. 

But from the comment above, I think we have a better solution that effectively does this without as much energy. 

---

So my recommendation boils down to something like:

1. Change the collision API to give the old location and the new location
2. Check the footprint as we do now. If the max cost of the line cost on the footprint is `< 253` or in collision, return respectively
3. If the footprint is `253` that means we're close to collision and its worth doing a bit more. We can interpolate from original_footprint -> new_footprint at a set resolution (say 10x) in a loop and add cells to a `std::set`
4. Once we completed the set, we iterate through them and check those for collision

Since we do that interpolation, that covers all the volumetric cells that would otherwise be encapsulated by the new footprint, plus some additional ones that would be colliding in transition due to the outward-most corners. It also removes redundant cell checks since we can also see if any in the `sweep_set` match values in the `new_footprint_set` so we only check the costmap for unique values without any duplication. 

I think its in your best interest to actually implement this in this case -- your solution misses those transition collisions when you have a small turning radius with big angular changes. If the volume check was an important fix for you, I think this is equally an important area that your current solution misses that you'd want

---

> I think that's pretty common so I'm surprised that issue was never reported in the past

Usually I do some filtering over the data before going into the costmap to remove those. We also have the salt and pepper denoising costmap layer now as well (though I haven't personally used it).

**Comment by tonynajjar** [comment]
I've exported the two findings of this issue into 2 tickets: https://github.com/ros-navigation/navigation2/issues/5329 and https://github.com/ros-navigation/navigation2/issues/5330. Closing this ticket in favor of those.
